feature: Configurable Agents As Collaborators

JediKev opened the pull request:
This adds a new setting called "Disable Agent Collaborators" that allows Admins to globally configure wether or not to allow Agents to be added as Collaborators via any communication on tickets. In some use cases Agents do not want to be added as Collaborators on tickets to avoid an overload of emails/alerts, etc. This adds a new checkbox labeled `Disable Agent Collaborators` to **Admin Panel > Settings > Agents** that if Enabled/Checked will automatically Disable Agents that are added as Collaborators. This disables the participant alerts for the Agents whilst giving other Agents/Admins the ability to enable them at a later date instead of having to manually add them back. If an Agent tries adding another Agent as a Collaborator via the Manage Collaborators modal it will also Disable the Agent Collaborator.

JediKev committed:
feature: Configurable Agents As Collaborators

This adds a new setting called "Disable Agent Collaborators" that allows Admins
to globally configure wether or not to allow Agents to be added as
Collaborators via any communication on tickets. In some use cases Agents do
not want to be added as Collaborators on tickets to avoid an overload of
emails/alerts, etc. This adds a new checkbox labeled `Disable Agent
Collaborators` to **Admin Panel > Settings > Agents** that if
Enabled/Checked will automatically Disable Agents that are added as
Collaborators. This disables the participant alerts for the Agents whilst
giving other Agents/Admins the ability to enable them at a later date
instead of having to manually add them back. If an Agent tries adding
another Agent as a Collaborator via the Manage Collaborators modal it will
also Disable the Agent Collaborator.

JediKev committed:
issue: Set Ticket Status For addCollaborator()

This addresses an issue where `Thread::getObject()` is not available for
`Thread::addCollaborator()` as the Ticket object is missing a Status
(incomplete). This sets the default Ticket Status (if none) so that the
Ticket object is complete and can be accessible via
`Thread::addCollaborator()`.

aydreeihn commented:
@JediKev We need to add an audit record for the thread event that says a collaborator was removed. I did this quick mockup to try to get rid of some of the redundancy in logging collaborator stuff. We can try to possibly think of better names for the variables, but here's what I put together:

```
diff --git a/include/class.thread.php b/include/class.thread.php
index 67e71caa..5af9e8a3 100644
--- a/include/class.thread.php
+++ b/include/class.thread.php
@@ -181,11 +181,8 @@ implements Searchable {
                 && Staff::lookup($user->getDefaultEmailAddress())) {
             if (!isset($vars['collab_del_event'])
                     || $vars['collab_del_event']) {
-                $this->getEvents()->log($this->getObject(),
-                    'collab',
-                    array('del' => array($user->getId() => array($user->getName()->getOriginal()))),
-                    'SYSTEM'
-                );
+                $vars['del'] = true;
+                $this->logCollaboratorEvents($user, $vars);
             } else
                 $errors['err'] = __('Agents cannot be added as Collaborators');
             return null;
@@ -200,20 +197,8 @@ implements Searchable {
         $this->_collaborators = null;
 
         if ($event) {
-          $this->getEvents()->log($this->getObject(),
-              'collab',
-              array('add' => array($user->getId() => array(
-                      'name' => $user->getName()->getOriginal(),
-                      'src' => @$vars['source'],
-                  ))
-              )
-          );
-
-          $type = array('type' => 'collab', 'add' => array($user->getId() => array(
-                  'name' => $user->getName()->name,
-                  'src' => @$vars['source'],
-              )));
-          Signal::send('object.created', $this->getObject(), $type);
+          $vars['add'] = true;
+          $this->logCollaboratorEvents($user, $vars);
         }
 
 
@@ -233,15 +218,7 @@ implements Searchable {
                         && ($c->getThreadId() == $this->getId())
                         && $c->delete())
                      $collabs[] = $c;
-
-                 $this->getEvents()->log($this->getObject(), 'collab', array(
-                     'del' => array($c->user_id => array('name' => $c->getName()->getOriginal()))
-                 ));
-                 $type = array('type' => 'collab', 'del' => array($c->user_id => array(
-                         'name' => $c->getName()->getOriginal(),
-                         'src' => @$vars['source'],
-                     )));
-                 Signal::send('object.deleted', $this->getObject(), $type);
+                 $this->logCollaboratorEvents($c, $vars);
             }
         }
 
@@ -284,6 +261,22 @@ implements Searchable {
         return true;
     }
 
+    function logCollaboratorEvents($collaborator, $vars) {
+        $name = $collaborator->getName()->getOriginal();
+        $userId = (get_class($collaborator) == 'User') ? $collaborator->getId() : $collaborator->user_id;
+        $action = $vars['del'] ? 'object.deleted' : 'object.created';
+        $addDel = $vars['del'] ? 'del' : 'add';
+
+        $this->getEvents()->log($this->getObject(), 'collab', array(
+            $addDel => array($userId => array('name' => $name))
+        ));
+        $type = array('type' => 'collab', $addDel => array($userId => array(
+                'name' => $name,
+                'src' => @$vars['source'],
+            )));
+        Signal::send($action, $this->getObject(), $type);
+    }
+
     //UserList of participants (collaborators)
     function getParticipants() {
```

JediKev commented:
@aydreeihn I updated this pull with the changes in a separate commit for better tracking. 👍

JediKev committed:
collab: Refuse To Add Owner As Collaborator

This adds a new `if()` statement to `Thread::addCollaborator()` that checks
the current Ticket Owner's ID to the User's ID that is being added as a
Collaborator. If the IDs match then we will refuse to add the Owner as a
Collaborator as Users are not supposed to be a Collaborator on Tickets they
own.

JediKev committed:
collab: Consolidate Collaborator Events

This adds a new method called `logCollaboratorEvents()` that logs any type
of collaborator event (del, add, etc.) and sends a signal for Audit Logs.
This helps consolidate all Collab Event logging in class Thread to make the
code cleaner and reduce redundancy.

JediKev commented:
Updated 👍
- Instead of ignoring agents and adding a thread event, we now disable them
- Implements `// FIXME: Refuse to add ticket owner??` - we now refuse to add the owner as a collab (just like fetching, etc.)

protich commented:
The tool tip message and variable used need to be updated inline to the new implementation.

JediKev commented:
@protich Updated 👍
